src/cpp/core/system/PosixSystem.cpp

```diff
@@ -533,9 +539,89 @@ Error closeFileDescriptorsFrom(int fdStart)
 
    return Success();
 }
+#else
+// read the file descriptors from a virtual directory listing,
+// iteratore over them and close - much faster than the above method
+Error closeFileDescriptorsFrom(int fdStart)
+{
+   std::vector<unsigned int> fds;
+   Error error = getOpenFds(&fds);
+   if (error)
+      return error;
+
+   for (int fd : fds)
+   {
+      if (fd >= fdStart)
+      {
+         if (::close(fd) < 0 && errno != EBADF)
+            return systemError(errno, ERROR_LOCATION);
+      }
+   }
+
+   return Success();
+}
+#endif
 
 } // anonymous namespace
 
+Error getOpenFds(std::vector<unsigned int>* pFds)
+{
+   return getOpenFds(getpid(), pFds);
+}
+
+#ifndef __APPLE__
+Error getOpenFds(pid_t pid, std::vector<unsigned int>* pFds)
+{
+   std::string pidStr = safe_convert::numberToString(pid);
+   boost::format fmt("/proc/%1%/fd");
+   FilePath fdDir = FilePath(boost::str(fmt % pidStr));
+
+   std::vector<FilePath> children;
+   Error error = fdDir.children(&children);
+   if (error)
+      return error;
+
+   for (const FilePath& child : children)
+   {
+      try
+      {
+         unsigned int fd = boost::lexical_cast<unsigned int>(child.filename());
+         pFds->push_back(fd);
+      }
+      catch (boost::bad_lexical_cast& err)
+      {
+         LOG_ERROR(systemError(boost::system::errc::invalid_argument, err.what(), ERROR_LOCATION));
+      }
+   }
+
+   return Success();
+}
+#else
+Error getOpenFds(pid_t pid, std::vector<unsigned int> *pFds)
+{
+   // get size of the buffer needed to hold the list of fds
+   int bufferSize = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, 0, 0);
+   if (bufferSize == -1)
+      return systemError(errno, ERROR_LOCATION);
+
+   // get the list of open fds
+   struct proc_fdinfo* procFdInfo = (struct proc_fdinfo*)malloc(bufferSize);
```

Comment:
nit: prefer `static_cast` of void pointers

Reply:
Fixed.